draft-kazuho-httpbis-priority.md

```diff
@@ -98,6 +98,70 @@ The terms sh-token and sh-boolean are imported from
 Example HTTP requests and responses use the HTTP/2-style formatting from
 {{?RFC7540}}.
 
+# Negotiating Priorities
+
+The HTTP/2 specification defines a priority scheme in [RFC7540],
+Section 5.3, which some implementers have opted not to fully support.
+The lack of signalling about the status of the implementation has
+caused several implementations to implement heuristics to detect when
+the clients they are connected to do not support priorities as
+defined and take steps to compensate for that.
+
+The intent of this negotiation is to provide a signalling mechanism
+for each peer to communicate which, if any, priority schemes are
+supported, as well as the server's ranked preference.
+
+For both HTTP/2 and HTTP/3, either peer's SETTINGS may arrive first,
+so any negotiation must be unilateral and not rely upon receiving
+the peer's SETTINGS value.
+
+Implementations are likely to only use one prioritization scheme at
+once, and may be unable to change the scheme once established, so the
+setting MUST be sent prior to the first request if it is ever sent.
+In HTTP/3, SETTINGS may arrive after the first request even if
+they are sent first.  In order to avoid the server incorrectly
+choosing a priority scheme in HTTP/3, the client SHOULD only send
+priority information prior to receiving the server's SETTINGS if
+it only offers a single scheme.  If it offers multiple schemes,
+the client SHOULD delay sending priority information until
+after it knows what the server supports, or risk the server
+choosing a sub-optimal scheme.
```

Comment:
Wondering if this SHOULD should be a MUST... if not, we might want to specify what a server is to do when it receives multiple, conflicting indicators (i.e., with this definition, it will start using the first one it sees, thinking it's the only one the client supports).

Reply:
I think the tradeoff is reasonable given the constraints. 

Trying to join up lists over multiple SETTINGS frames is not something I would like to implement.

Using a Frame allows as much space as needed, I spaced it out in an unshared doc, but I dont like the additional work it adds and the cost of losing support for remembered SETTINGSs

Reply:
IMO, there are two problems with the text.

One is that it might be too restrictive. We do not know what other priority schemes we'd see in the future. Consider the case where we'd have a frame-based prioritization scheme that is dedicated to video requests, while the other streams would be using the Priority headers. In such case, there is no reason to prohibit either of the two from being sent until the client learns that the SETTINGS have been received.

The other problem is that H3 does not define a way to acknowledge the receipt of SETTINGS at the application-protocol level. It is my understanding that H3 (including QPACK) is designed to have it's own ACKs for signaling the receipt, rather than relying on transport-level ACKs. This normative text violates that principle.

Considering these aspects, I think that we should paraphrase this these sentences discussing HTTP/3 to something like: _In HTTP/3, SETTINGS may arrive after the first request even if they are sent first. Future specifications that define alternative prioritization schemes SHOULD define how the server SHOULD act it receives stream-level priority signals prior to receiving the SETTINGS frame_.